bing_bert/nvidia/modelingpreln_layerdrop.py

```diff
@@ -0,0 +1,1662 @@
+                attn_dropout_ratio=config.attention_probs_dropout_prob,
+                hidden_dropout_ratio=config.hidden_dropout_prob,
+                num_hidden_layers=config.num_hidden_layers,
+                initializer_range=config.initializer_range,
+                local_rank=args.local_rank
+                if hasattr(args, 'local_rank') else -1,
+                seed=args.seed,
+                fp16=ds_config.fp16_enabled,
+                pre_layer_norm=True,
+                attn_dropout_checkpoint=args.attention_dropout_checkpoint,
+                normalize_invertible=args.normalize_invertible,
+                gelu_checkpoint=args.gelu_checkpoint,
+                stochastic_mode=args.stochastic_mode)
+
+            self.layer = nn.ModuleList([
+                copy.deepcopy(DeepSpeedTransformerLayer(i, cuda_config))
+                for i in range(config.num_hidden_layers)
+            ])
+        else:
+            layer = BertLayer(config)
+            if sparse_attention_config is not None:
+                from deepspeed.ops.sparse_attention import BertSparseSelfAttention
+
+                layer.attention.self = BertSparseSelfAttention(
+                    config, sparsity_config=sparse_attention_config)
+
+            self.layer = nn.ModuleList([
+                copy.deepcopy(layer) for _ in range(config.num_hidden_layers)
+            ])
+
+    # def forward(self, hidden_states, attention_mask, output_all_encoded_layers=True):
+    #     all_encoder_layers = []
+    #     for layer_module in self.layer:
+    #         hidden_states = layer_module(hidden_states, attention_mask)
+    #         if output_all_encoded_layers:
+    #             all_encoder_layers.append(hidden_states)
+    #     if not output_all_encoded_layers:
+    #         all_encoder_layers.append(hidden_states)
+    #     return all_encoder_layers
+    def forward(self,
+                hidden_states,
+                attention_mask,
+                output_all_encoded_layers=True,
+                checkpoint_activations=False,
+                progressive_layer_drop=False,
+                theta=0.5):
+        all_encoder_layers = []
+
+        def custom(start, end):
+            def custom_forward(*inputs):
+                layers = self.layer[start:end]
+                x_ = inputs[0]
+                for layer in layers:
+                    x_ = layer(x_, inputs[1])
+                return x_
+
+            return custom_forward
+
+        if checkpoint_activations:
+            l = 0
+            num_layers = len(self.layer)
+            chunk_length = math.ceil(math.sqrt(num_layers))
+            while l < num_layers:
+                hidden_states = checkpoint.checkpoint(
+                    custom(l, l + chunk_length), hidden_states,
+                    attention_mask * 1)
+                l += chunk_length
+            # decoder layers
+        else:
+            if not progressive_layer_drop:

```

Comment:
This large branch makes me wonder if we can just subclass `BertEncoder` and implement the custom `forward()` ?

Reply:
That's a great point. The users currently can do a comparison between the modelingpreln_layerdrop and modelingpreln to see the changes. I will do a refactoring and test of the code once the DSE changes have been done. For now, I wonder if we can use the separate modelingpre_layerdrop file, given that we have an internal deadline coming very soon.